### PR TITLE
Bump up to Go 1.21

### PR DESCRIPTION
**Pull request by SamYuan1990**
as fix for #934

### CODE REVIEW

**Comment by SamYuan1990** [comment]
ERROR: failed to solve: quay.io/sustainable_computing_io/kepler_builder:ubi-9-bcc-0.26-go1.21: quay.io/sustainable_computing_io/kepler_builder:ubi-9-bcc-0.26-go1.21: not found

@rootfs , if we want to bump up version to golang 1.21, we need to have base image.

**Comment by rootfs** [comment]
@marceloamaral @sunya-ch any comment on upgrading golang?

**Comment by jiere** [comment]
> ERROR: failed to solve: quay.io/sustainable_computing_io/kepler_builder:ubi-9-bcc-0.26-go1.21: quay.io/sustainable_computing_io/kepler_builder:ubi-9-bcc-0.26-go1.21: not found

It should be built from Dockerfile.bcc.builder(in next two releases, since bcc will be removed, the builder may also evolve to Dockerfile.builder?), we need to define a mechanism for such bump requirement to update the builder image, sounds like an ad-hoc github action.

**Comment by SamYuan1990** [comment]
Please allow me to response via mail before go to sleep today.
For image rename, we don’t cover in this PR, for this PR we just focus on golang bump up.
For image or base image I suppose we should keep it for a long time.
The totally remove of the image which means we end support for bcc based kepler.
Before we make it(including any new issue open or image refresh request to fix CVE) or the day comes.
We have to keep the base image, and ideally, refresh them.


| |
Sam Yuan
|
|
***@***.***
Thanks and Regards
|


---- Replied Message ----
| From | Huamin ***@***.***> |
| Date | 09/18/2023 23:36 |
| To | ***@***.***> |
| Cc | Sam ***@***.***> ,
***@***.***> |
| Subject | Re: [sustainable-computing-io/kepler] Bump up to Go 1.21 (PR #939) |

I am leaning to move this to post 0.6, wdyt @SamYuan1990

**Comment by sunya-ch** [comment]
I think it would be better to have the newly-proposed base image test first. It could be applied once we have change on the base image Docker file and relevant path.

**Comment by rootfs** [comment]
@sunya-ch agreed

Since the builder images are locally built and pushed. It is time to move them to a CI action so we can have builders on demand.

@SamYuan1990 would you like to create a GHA to build the builder image and base image? So we'll have these images first before we try the go 1.21
https://github.com/sustainable-computing-io/kepler/blob/main/build/Dockerfile.bcc.builder
https://github.com/sustainable-computing-io/kepler/blob/main/build/Dockerfile.libbpf.builder
https://github.com/sustainable-computing-io/kepler/blob/main/build/Dockerfile.bcc.base
https://github.com/sustainable-computing-io/kepler/blob/main/build/Dockerfile.libbpf.base

**Comment by sunya-ch** [comment]
> > @sunya-ch agreed
> > Since the builder images are locally built and pushed. It is time to move them to a CI action so we can have builders on demand.
> > @SamYuan1990 would you like to create a GHA to build the builder image and base image? So we'll have these images first before we try the go 1.21 https://github.com/sustainable-computing-io/kepler/blob/main/build/Dockerfile.bcc.builder https://github.com/sustainable-computing-io/kepler/blob/main/build/Dockerfile.libbpf.builder https://github.com/sustainable-computing-io/kepler/blob/main/build/Dockerfile.bcc.base https://github.com/sustainable-computing-io/kepler/blob/main/build/Dockerfile.libbpf.base
> 
> wait a min... does it mean, we need a new build to build base image. and which triggered by push event and daily run?

No, we should limit it to the change on file path that affect the base image such as Dockerfile.base.

**Comment by SamYuan1990** [comment]
we have nvidia based bcc image and nvidia based libbpf image for x86, arm and amd.
and
ubi based image for s390x
hence, considering with current header issue mentioned at https://github.com/sustainable-computing-io/kepler/pull/587#discussion_r1211813389 for bump up version to go 1.21
I am going to have nvidia based bcc image and nvidia based libbpf image for x86, arm and amd, as a single base image with buildx supported.
for s390, I suppose we need further consider it. cc + @jichenjc in loop here.

and then, at implement level.
1st consider make the base image build triggered by PR.
2nd consider make the base image build as daily build.

File: build/Dockerfile.bcc.builder [hunk 1]
```diff
@@ -7,4 +7,4 @@ LABEL name=kepler-builder
 RUN yum install -y kernel-devel make git gcc rpm-build systemd && \
     yum clean all -y 
 
-RUN curl -LO https://go.dev/dl/go1.18.10.linux-amd64.tar.gz; mkdir -p /usr/local; tar -C /usr/local -xvzf go1.18.10.linux-amd64.tar.gz; rm -f go1.18.10.linux-amd64.tar.gz
+RUN curl -LO https://go.dev/dl/go1.21.1.linux-amd64.tar.gz; mkdir -p /usr/local; tar -C /usr/local -xvzf go1.21.1.linux-amd64.tar.gz; rm -f go1.21.1.linux-amd64.tar.gz
```
Comment (rootfs) [inline build/Dockerfile.bcc.builder:10]: can you make this a separate PR? The new builder image is needed for all other upgrade.
